extmod/modure.c

```diff
@@ -253,8 +257,15 @@ MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(re_split_obj, 2, 3, re_split);
 
 #if MICROPY_PY_URE_SUB
 
-STATIC mp_obj_t re_sub_helper(mp_obj_t self_in, size_t n_args, const mp_obj_t *args) {
-    mp_obj_re_t *self = MP_OBJ_TO_PTR(self_in);
+STATIC mp_obj_t re_sub_helper(size_t n_args, const mp_obj_t *args) {
+    mp_obj_re_t *self;
+    if (mp_obj_is_str_or_bytes(args[0])) {
+        self = MP_OBJ_TO_PTR(mod_re_compile(1, args));
+    } else if (mp_obj_is_type(args[0], &re_type)) {
+        self = MP_OBJ_TO_PTR(args[0]);
+    } else {
+        mp_raise_TypeError(MP_ERROR_TEXT("not a re object"));
+    }
```

Comment:
I think it'd be simpler to switch the check around like this:
```c
if (mp_obj_is_type(args[0], &re_type)) {
     self = MP_OBJ_TO_PTR(args[0]);
} else {
    self = MP_OBJ_TO_PTR(mod_re_compile(1, args));
}
```
Then it's up to `mod_re_compile()` to raise an exception if the argument is not a string (which is how it works at the moment).